### ISSUE TITLE
Support Jakarta - tomcat 10/11

### ISSUE THREAD

**Issue by owenpmckenna** [psi-probe/psi-probe#2394]
Will psi probe be updated for Tomcat 10? I tried to mess with it myself (find/replace "javax" with "jakarta", basically), but it turns out, like, Sitemesh? Is dead? And other problems... Mostly dependency based, along with my terrible knowledge of Tomcat. So, all that to say, will psi probe be updated for Tomcat 10 in the near future? (I might possibly be an idiot about something, when I tried to install psi probe on my server and went to /probe I was shown a blank white screen, if that's a common problem or something) Thanks!

**Comment by hazendaz** [comment]
Its complicated.  There are too many libraries that still are javax based, all have to upgrade to jakarta entirely.  Additionally all legacy support here has to be dropped entirely.  I have the modules in play in the build already for jakarta.  I need to move to spring 6, spring security 6, etc.  That itself means drop all prior support as the project itself is spring based and can only be one thing.  Currently I'm the only person supporting this product and my use is basically zero as moved onto spring boot on all my projects, so finding time to try to figure out massive upgrades needed requires the large community of users to actually be involved.  The download stats are high but frankly no one helps.  I welcome all help but as you can see, just lots of issues, no pull requests :(

**Comment by hazendaz** [comment]
sitemesh is a blocker currently.  it will need to be pulled and maintained and put on jakarta before we can progress this at all.

**Comment by hazendaz** [comment]
I've re-established project and brought it considerably up-to-date over at https://github.com/hazendaz/sitemesh2.  There is still more work to do there to get a javax copy out as code is considerably with issues currently.  Expect to have that likely out next weekend and tested against current psi probe javax release.  If that all works out, that release will be cut along with tomcat 9.0.74 and spring patches to address current CVEs under javax.  After that will start working on getting a jakartaEE equivalent out.  Expecting few more weeks at least on that if not more.  No real ETA on any of this otherwise.

**Comment by hazendaz** [comment]
I've got equivalent versions of sitemesh now out for both javax and jakartaEE.  One user has already noted sitemesh worked out well for them on jakartaEE.  Plan to get releases out this weekend with psi probe unless I run into any further issues.

**Comment by hazendaz** [comment]
Tomcat 10.1 removed ServletFileUpload.  That is currently a blocking issue for progress.  Site mesh worked out so close to finish line.

**Comment by pizzi80** [comment]
What's the problem with upload?

https://www.baeldung.com/upload-file-servlet

**Comment by hazendaz** [comment]
code needs rewritten.  current project uses that class that no longer exists.

**Comment by pizzi80** [comment]
if you copy/paste that class from the old source of tomcat?

https://dlcdn.apache.org/tomcat/tomcat-9/v9.0.76/src/apache-tomcat-9.0.76-src.tar.gz

**Comment by hazendaz** [comment]
Thanks will take a look when I can.  If you have some time and don't mind raising PR that would be great.  At very least, thank you for the information.

**Comment by hazendaz** [comment]
Just an update here, I have split two development lines which are still the same currently 'master' is going to be 'jakarta' and a branch called 'javax' will be the javax namespace.  I've been doing a lot of code cleanup and have Renovate positioned to help having two distinct lines until end of 2024 when javax will go dormant.

I have read on why tomcat got rid of the fileupload logic there as it was not designed for usage in same way.  There is however a larger effort to fileupload 2 in the works so I expect but did not look that will be some solution.  I'm not yet certain how to address this but the suggestion here to just copy their code may be required for this project given its usage.

I am trying to get rid of as much deprecated code as possible, raise code coverage, cleanup issues within the code, etc.  So I expect a few more months likely before I push out jakarta.  I haven't provided out what I've done to anyone but will try to do so by year end to others can see the work in progress.  That will be on a separate branch as I keep rebasing it while trying to make sure javax and jakarta are both on equal footing as much as possible.

There are a number of things I have been trying to do over time in general so anyone reading this, if you look at my fork you can see where my head has been for years in some areas and either lack of time or experience on some of the migrations has held things up.  See https://github.com/hazendaz/psi-probe/.   Do ignore the jakarta branch there, that was an earlier effort to split the repo in one place which was a failed effort due to spring.  But generally other items kind of show what I'm thinking.  Such things as trying to upgrade spring security issues, get off legacy javascript, try to use webjars, get more loggers directly working, etc.  Any community help in those areas would be great.

**Comment by hazendaz** [comment]
Raised 'jarkarta-wip' branch here directly with work in progress.  Going to document some of that on a roadmap in discussions for clarity.

**Comment by hazendaz** [comment]
@pizzi80 I presume the code snippet you have would work with 'javax' namespace too.  We are still fixing up some issues in the two modules that matter there.  It would be great if you could take the work you have on that other repo and adopt that to work with our setup on the 'javax' stream if its easily enough to get it working there and submit us a pull request.  If that works out as I have folks heavily testing that area now, it would make release of jakarta happen really fast.

**Comment by pizzi80** [comment]
Hi @hazendaz,

I have no more project on the old "javax" servlet api ...

Due to this, at the moment I'm not using anymore psi probe

I'm waiting for "jakarta" version of psi-probe ! :crossed_fingers: 

At this moment I'm very busy and I don't use Spring 
so to me it would require a lot of time to contribute to this project ... :disappointed: 

In 2023 in my spare time I've created an experimental 
Faces+CDI version of psi probe ... but it's not complete

I think that If you use Faces + CDI with Java 21 
you could rewrite this project with 1/3 of the code ... 
... It's probably a one month work ... 

I'm really sorry that I could not contribute :(

**Comment by hazendaz** [comment]
@pizzi80 No worries thanks for the insight and suggested directions.  Appreciate it.

**Comment by hazendaz** [comment]
Tomcat rewrote and pushed a pretty breaking change back to tomcat 9 that effectively showed the light here.  The changes I had to make to support tomcat 9 actually work with tomcat 10/11.  I need to do testing still but expect to have jakarta support out soon.

**Comment by tjay** [comment]
Hi @hazendaz, can you link this "pretty breaking change". Thanks!

**Comment by hazendaz** [comment]
Not sure what you mean.  Jakarta support hasn't been released yet.

Sent from my Verizon, Samsung Galaxy smartphone
Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by tjay** [comment]
Sorry, but maybe I just misunderstood you: but I thought you were referring to a specific change in the Tomcat project that recently occurred that will help you fix this issue

**Comment by hazendaz** [comment]
The changes to that have already been applied to this repo.  In fact, the tomcat9 based ones were already released.  Its internal and noted in the change log.  Had to do with how file uploads work.  That same change was the only blocking issue with tomcat 10 migration here so that work is done.  The only delay at this point is testing effort involved.

**Comment by hazendaz** [comment]
@tjay If it helps to see what tomcat did, you can see here https://github.com/apache/tomcat/commit/9ba3b0c57fff311cd898716db19e508183545b28#commitcomment-141608023.  The class in question was already done from future tomcat versions so it simply forced the changes to tomcat 9.  As someone asked there, they didn't use that class and users should not use internal apis.  However, psi probe requires internal apis to work and thus the latest release for tomcat 9 from here was adjusted for that.  From a user stand point that shouldn't mean anything.

**Comment by hazendaz** [comment]
Also, any old tomcat 8.5 or 9 will work with the change noted as its no longer using tomcat internals but more directly using the code tomcat had (was 2 classes and now in one class here for the parts needed only).

**Comment by hazendaz** [comment]
Released psi-probe 5.0.0 for jakarta support.

**Comment by pizzi80** [comment]
This is a very good news! :champagne:
